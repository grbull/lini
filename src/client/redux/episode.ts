/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type EpisodeState = {
  isLoading: boolean;
  error?: string;
  data?: EpisodeDto;
};

const initialState: EpisodeState = {
  isLoading: false,
};

export const get = createAsyncThunk<EpisodeDto, number>('episode/get', (id) =>
  api.episode.get(id)
);

export const episode = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const episodeActions = { ...episode.actions, get };
