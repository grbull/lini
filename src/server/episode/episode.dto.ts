import { ShowDto } from '../show/show.dto';

export class EpisodeDto {
  id!: string;
  name!: string;
  season!: number;
  number!: number;
  type!: string;
  airdate?: string;
  airtime?: string;
  airstamp?: string;
  runtime?: number;
  imageMedium?: string;
  imageOriginal?: string;
  summary?: string;
  show!: ShowDto;
}

export class ScheduleDto {
  future!: EpisodeDto[];
  past!: EpisodeDto[];
}
