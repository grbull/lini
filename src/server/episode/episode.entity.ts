import { Column, Entity, ManyToOne } from 'typeorm';

import { ShowEntity } from '../show/show.entity';

@Entity('episode')
export class EpisodeEntity {
  @Column('bigint', { primary: true })
  id!: string;

  @Column('text')
  name!: string;

  @Column('integer')
  season!: number;

  @Column('integer')
  number!: number;

  @Column('text')
  type!: string;

  @Column('date', { nullable: true })
  airdate!: string | null;

  @Column('time', { nullable: true })
  airtime!: string | null;

  @Column('timestamptz', { nullable: true })
  airstamp!: string | null;

  @Column('integer', { nullable: true })
  runtime!: number | null;

  @Column('text', { nullable: true })
  imageMedium!: string | null;

  @Column('text', { nullable: true })
  imageOriginal!: string | null;

  @Column('text', { nullable: true })
  summary!: string | null;

  @ManyToOne(() => ShowEntity, (show) => show.episodes)
  show!: number;
}
