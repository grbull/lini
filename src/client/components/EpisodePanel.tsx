import cn from 'classnames';
import React, { ReactElement } from 'react';
import { Link } from 'react-router-dom';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { dateFromNow } from '../utils/date';
import { formatEpisodeCode } from '../utils/formatEpisodeCode';
import { imageShow } from '../utils/imageShow';
import { EpisodePanelImage } from './EpisodePanelImage';

interface Props {
  episodes: EpisodeDto[];
}

export function EpisodePanel({ episodes }: Props): ReactElement {
  return (
    <div
      className={cn('w-full', 'overflow-x-auto', 'whitespace-nowrap', 'px-1')}
    >
      {episodes.map((episode) => (
        <Link
          className={cn('inline-block', 'w-1/3', 'p-1')}
          key={episode.id}
          to={`/episode/${episode.id}`}
        >
          <EpisodePanelImage
            alt={episode.show.name}
            imageURL={imageShow(episode.show.imageMedium)}
          />
          <div className={cn('h-18')}>
            <span
              className={cn(
                'block',
                'whitespace-nowrap',
                'overflow-hidden',
                'overflow-ellipsis',
                'font-bold',
                'mt-1'
              )}
            >
              {episode.show.name}
            </span>
            <span className={cn('block')}>
              {formatEpisodeCode(episode.season, episode.number)}
            </span>
            <time className={cn('block')} dateTime={episode.airstamp}>
              {episode.airstamp ? dateFromNow(episode.airstamp) : 'N/A'}
            </time>
          </div>
        </Link>
      ))}
    </div>
  );
}
