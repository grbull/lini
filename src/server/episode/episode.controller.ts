import {
  Controller,
  Get,
  Param,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';

import { User } from '../common/decorators/user.decorator';
import { AuthGuard } from '../common/guards/auth.guard';
import { LoggingInterceptor } from '../common/interceptors/logging.interceptor';
import { UserEntity } from '../user/user.entity';
import { EpisodeEntity } from './episode.entity';
import { EpisodeService } from './episode.service';

@Controller('episode')
@UseGuards(AuthGuard)
@UseInterceptors(LoggingInterceptor)
export class EpisodeController {
  constructor(private readonly episodeService: EpisodeService) {}

  @Get('schedule')
  public async schedule(
    @User() user: UserEntity
  ): Promise<{ future: EpisodeEntity[]; past: EpisodeEntity[] }> {
    const future = await this.episodeService.getFuture(user);
    const past = await this.episodeService.getPast(user);
    return { future, past };
  }

  @Get(':id')
  public getOne(@Param('id') id: number): Promise<EpisodeEntity> {
    return this.episodeService.getOne(id);
  }
}
