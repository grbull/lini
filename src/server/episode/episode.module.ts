import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { LoggerModule } from '../logger/logger.module';
import { SubscriptionModule } from '../subscription/subscription.module';
import { EpisodeController } from './episode.controller';
import { EpisodeEntity } from './episode.entity';
import { EpisodeService } from './episode.service';

@Module({
  imports: [
    LoggerModule,
    TypeOrmModule.forFeature([EpisodeEntity]),
    SubscriptionModule,
  ],
  providers: [EpisodeService],
  controllers: [EpisodeController],
  exports: [EpisodeService],
})
export class EpisodeModule {}
