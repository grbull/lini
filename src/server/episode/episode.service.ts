import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Between, Repository } from 'typeorm';

import { LoggerService } from '../logger/logger.service';
import { SubscriptionService } from '../subscription/subscription.service';
import { TvMazeEpisode } from '../tvmaze/tvmaze.types';
import { tvMazeEpisodeToEntity } from '../tvmaze/tvmaze.utils';
import { UserEntity } from '../user/user.entity';
import { EpisodeEntity } from './episode.entity';

@Injectable()
export class EpisodeService {
  constructor(
    private readonly loggerService: LoggerService,
    @InjectRepository(EpisodeEntity)
    private readonly episodeRepository: Repository<EpisodeEntity>,
    private readonly subscriptionService: SubscriptionService
  ) {
    this.loggerService.setContext('EpisodeService');
  }

  public async getOne(id: number): Promise<EpisodeEntity> {
    try {
      return await this.episodeRepository
        .createQueryBuilder('episode')
        .leftJoinAndSelect('episode.show', 'show')
        .where('episode.id = :id', { id })
        .getOneOrFail();
    } catch {
      throw new NotFoundException();
    }
  }

  // public getByShow(id: number): Promise<EpisodeEntity[]> {
  //   return this.episodeRepository
  //     .createQueryBuilder('episode')
  //     .where('episode.show.id = :id', { id })
  //     .orderBy('episode.airdate', 'ASC')
  //     .getMany();
  // }

  // public async getUpcoming(user: UserEntity): Promise<EpisodeEntity[]> {
  //   const subscriptions = await this.subscriptionService.getAll(user);
  //   let upcomingEpisodes: EpisodeEntity[] = [];

  //   const oneMonthAhead = new Date();
  //   oneMonthAhead.setMonth(new Date().getMonth() + 1);

  //   for (const key of subscriptions) {
  //     const episodes = await this.episodeRepository.find({
  //       where: {
  //         // between now and 1 month ahead
  //         // airstamp: Raw(
  //         //   (alias) => `${alias} < '${oneMonthAhead.toISOString()}'`
  //         // ),
  //         airstamp: Between(new Date(), oneMonthAhead),
  //         show: key.show,
  //       },
  //     });
  //     console.log(episodes);
  //     if (episodes) {
  //       upcomingEpisodes = [...upcomingEpisodes, ...episodes];
  //     }
  //   }
  //   return upcomingEpisodes;
  // }

  public async getFuture(user: UserEntity): Promise<EpisodeEntity[]> {
    const subscriptions = await this.subscriptionService.getAll(user);

    const oneMonthAhead = new Date();
    oneMonthAhead.setMonth(new Date().getMonth() + 1);

    const findOptions = subscriptions.map((subscription) => {
      return {
        airstamp: Between(new Date(), oneMonthAhead),
        show: subscription.show,
      };
    });

    return this.episodeRepository.find({
      where: findOptions,
      order: {
        airstamp: 'ASC',
        number: 'ASC',
      },
      relations: ['show'],
    });
  }

  public async getPast(user: UserEntity): Promise<EpisodeEntity[]> {
    const subscriptions = await this.subscriptionService.getAll(user);

    const oneMonthBehind = new Date();
    oneMonthBehind.setMonth(new Date().getMonth() - 1);

    const findOptions = subscriptions.map((subscription) => {
      return {
        airstamp: Between(oneMonthBehind, new Date()),
        show: subscription.show,
      };
    });

    return this.episodeRepository.find({
      where: findOptions,
      order: {
        airstamp: 'DESC',
        number: 'DESC',
      },
      relations: ['show'],
    });
  }

  // The belowe is used for seeding and updating
  public createOrUpdateMany(
    episodes: TvMazeEpisode[],
    showId: number
  ): Promise<EpisodeEntity[]> {
    const episodeEntities = episodes.map((episode) =>
      tvMazeEpisodeToEntity(episode, showId)
    );

    return this.episodeRepository.save(episodeEntities, { chunk: 1000 });
  }

  // used by notification service
  public findByTime(
    showId: number,
    startTime: Date,
    endTime: Date
  ): Promise<EpisodeEntity[]> {
    return this.episodeRepository.find({
      where: { show: showId, airstamp: Between(startTime, endTime) },
    });
  }
}
