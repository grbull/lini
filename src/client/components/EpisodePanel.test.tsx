/**
 * @jest-environment jsdom
 */
/* eslint-disable init-declarations */

import '@testing-library/jest-dom/extend-expect';

import { render } from '@testing-library/react';
import { createMemoryHistory, MemoryHistory } from 'history';
import React from 'react';
import { Router } from 'react-router-dom';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { EpisodePanel } from './EpisodePanel';

const episodes: EpisodeDto[] = [
  {
    id: '2095477',
    name: 'Episode 5',
    season: 1,
    number: 5,
    type: 'regular',
    airdate: '2021-07-07',
    airtime: null,
    airstamp: '2021-07-07T12:00:00.000Z',
    runtime: null,
    imageMedium: null,
    imageOriginal: null,
    summary: null,
    show: {
      id: 41007,
      name: 'Loki',
      type: 'Scripted',
      language: 'English',
      genres: ['Action', 'Adventure', 'Science-Fiction'],
      status: 'Running',
      runtime: null,
      runtimeAverage: 48,
      datePremiered: '2021-06-09',
      officialSite: 'https://www.marvel.com/tv-shows/loki/1',
      scheduleTime: null,
      scheduleDays: ['Wednesday'],
      rating: 8.2,
      imageMedium: '/uploads/images/medium_portrait/320/801227.jpg',
      imageOriginal: '/uploads/images/original_untouched/320/801227.jpg',
      summary:
        'Loki follows the trickster and shape-shifter god who pops up throughout human history as an unlikely influencer on historical events.',
      network: null,
      webChannel: null,
    },
  },
];

describe('EpisodePanel Component', () => {
  let history: MemoryHistory<unknown>;
  Date.now.bind(global.Date);
  const dateNowStub = jest.fn(() => 1625584251217);
  global.Date.now = dateNowStub;

  beforeEach(() => {
    history = createMemoryHistory();
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <Router history={history}>
        <EpisodePanel episodes={episodes} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render episodes without an airstamp', () => {
    const { getByText } = render(
      <Router history={history}>
        <EpisodePanel episodes={[{ ...episodes[0], airstamp: null }]} />
      </Router>
    );
    expect(getByText('N/A')).toBeTruthy();
  });
});
