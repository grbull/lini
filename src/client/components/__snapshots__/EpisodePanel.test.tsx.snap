// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodePanel Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full overflow-x-auto whitespace-nowrap px-1"
  >
    <a
      class="inline-block w-1/3 sm:w-1/4 md:w-1/5 lg:w-1/6 p-1"
      href="/episode/2095477"
    >
      <div
        class="h-0 relative rounded-sm"
        style="padding-top: 140.47619047619045%;"
      >
        <div
          class="animate-pulse bg-gray-700 absolute top-0 left-0 w-full h-full bg-gray-800"
        />
        <img
          alt="Loki"
          class="absolute top-0 left-0 w-full h-full hidden"
          src="undefined/uploads/images/medium_portrait/320/801227.jpg"
        />
      </div>
      <div
        class="h-18"
      >
        <span
          class="block whitespace-nowrap overflow-hidden overflow-ellipsis text-lg font-semibold mt-1"
        >
          Loki
        </span>
        <span
          class="block text-sm"
        >
          S01E05
        </span>
        <time
          class="block text-sm font-light"
          datetime="2021-07-07T12:00:00.000Z"
        >
          in 20 hours
        </time>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
